### PR TITLE
Implement cored and truncated nfw profile 

### PR DESCRIPTION
**Pull request by dangilman**
I add an NFW-like mass profile with a core radius and a truncation radius. I got analytic expressions for the projected mass and the integral of the projected mass by moving the square inside the parenthesis in the denominator, i.e. (1+x^2) instead of (1+x)^2. The core is implemented through a term (r_c^2 + r^2)^1/2, and the profile drops like 1/r^5 outside the truncation radius.

### CODE REVIEW

**Review by sibirrer — CHANGES_REQUESTED** [review]
Looks great! I have only some minor documentation requests.
I guess this is a new profile and perhaps you can leave a reference to a forthcoming publication or so?

File: lenstronomy/LensModel/Profiles/nfw_core_truncated.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        beta = r_core / Rs
+        tau = r_trunc / Rs
+        gx = self._g(x, beta, tau)
+        m_2d = 4 * rho0 * Rs * R ** 2 * gx / x ** 2 * np.pi
+        return m_2d
+
+    def nfwAlpha(self, R, Rs, rho0, r_core, r_trunc, ax_x, ax_y):
+        """
+
+        deflection angel of NFW profile (times Sigma_crit D_OL) along the projection to coordinate 'axis'
```
[the remark this answers is not in the document]
Comment (dangilman) [reply lenstronomy/LensModel/Profiles/nfw_core_truncated.py]: fixed

File: lenstronomy/LensModel/Profiles/nfw_core_truncated.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+        """
+        R = np.maximum(R, 0.00000001)
+        x = R / Rs
+        beta = r_core / Rs
+        tau = r_trunc / Rs
+        gx = self._g(x, beta, tau)
+        m_2d = 4 * rho0 * Rs * R ** 2 * gx / x ** 2 * np.pi
+        return m_2d
+
+    def nfwAlpha(self, R, Rs, rho0, r_core, r_trunc, ax_x, ax_y):
```
[the remark this answers is not in the document]
Comment (dangilman) [reply lenstronomy/LensModel/Profiles/nfw_core_truncated.py]: done

File: lenstronomy/LensModel/Profiles/nfw_core_truncated.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+class TNFWC(LensProfileBase):
+    """
+
+    This class contains an pseudo NFW profile with a core radius and a truncation radius. The density in 3D is given by
+
+    .. math::
+        \\rho(r) = \\frac{\\rho_0 r_s^\3}{\left(r^2+r_c^2\right)^{1/2} \left(r_s^2+r^2\right)} \left(\frac{r_t^2}{r^2+r_t^2}\right)
+
+    For r_c -> 0 and r_t -> infinity, this profile reduce to something like an NFW profile with the squared term in
```
[the remark this answers is not in the document]
Comment (dangilman) [reply lenstronomy/LensModel/Profiles/nfw_core_truncated.py]: So I'm not actually sure how to format it, but I tried. Can you have a look?

File: lenstronomy/LensModel/Profiles/nfw_core_truncated.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
+__author__ = 'dgilman'
+
+# this file contains a class to compute lensing proprerties of a pseudo Navaro-Frenk-White profile with a core and truncation
+# radius
+import numpy as np
+from lenstronomy.LensModel.Profiles.base_profile import LensProfileBase
+from scipy.integrate import quad
+
+__all__ = ['TNFWC']
+
```
[the remark this answers is not in the document]
Comment (dangilman) [reply lenstronomy/LensModel/Profiles/nfw_core_truncated.py:10]: done

**Comment by dangilman** [comment]
by the way, in this pull request I edited the MultiPatchReconstruction class to fix an issue I encountered where the code crashes when the pixel coordinate corresponding to each patch is not quite an integer. I fixed it by forcing them to be ints

File: lenstronomy/Plots/model_band_plot.py
Comment (dangilman) [inline lenstronomy/Plots/model_band_plot.py:1]: the change here fixes an issue where sometimes the image labels don't show

File: lenstronomy/LensModel/Profiles/nfw_core_truncated.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+class TNFWC(LensProfileBase):
+    """
+
+    This class contains an pseudo NFW profile with a core radius and a truncation radius. The density in 3D is given by
+
+    .. math::
+        \\rho(r) = \\frac{\\rho_0 r_s^\3}{\left(r^2+r_c^2\right)^{1/2} \left(r_s^2+r^2\right)} \left(\frac{r_t^2}{r^2+r_t^2}\right)
+
+    For
```
Comment (sibirrer) [inline lenstronomy/LensModel/Profiles/nfw_core_truncated.py:20]: I was more thinking of " For :math:`\\r_c \\rightarrow 0` ..." In that way you can embed lateX in text and not as a formula

Comment (dangilman) [reply lenstronomy/LensModel/Profiles/nfw_core_truncated.py]: ok, how about now?

Comment (sibirrer) [reply lenstronomy/LensModel/Profiles/nfw_core_truncated.py]: oh, you need to keep the parentheses, they got converted here unfortunately. See e.g. https://github.com/lenstronomy/lenstronomy/blob/b9d03df087c846566eb9be777cc2e9a0a7e8a751/lenstronomy/LensModel/Profiles/epl.py#L21

**Comment by sibirrer** [comment]
Thank you @dangilman ! According to coveralls, there are 8 lines missing to be tested in the new profile (unfortunately, it does not show up). Perhaps you can run 
$make coverage
locally to see the full coverage report

**Comment by dangilman** [comment]
> Thank you @dangilman ! According to coveralls, there are 8 lines missing to be tested in the new profile (unfortunately, it does not show up). Perhaps you can run $make coverage locally to see the full coverage report

ok, what command do I enter in the terminal to do this?

**Comment by sibirrer** [comment]
> > Thank you @dangilman ! According to coveralls, there are 8 lines missing to be tested in the new profile (unfortunately, it does not show up). Perhaps you can run $make coverage locally to see the full coverage report
> 
> ok, what command do I enter in the terminal to do this?

`make coverage`

this will execute these lines in the make file: https://github.com/lenstronomy/lenstronomy/blob/b9d03df087c846566eb9be777cc2e9a0a7e8a751/Makefile#L36

**Comment by dangilman** [comment]
> I got this error... 

Screen Shot 2023-08-11 at 5 24 06 PM

**Comment by dangilman** [comment]
I added some tests trying to guess which lines weren't getting hit... we'll see

**Comment by sibirrer** [comment]
aside from the test crashing, if you install
pip install coverage
the command should run.

**Pull request merged**